src/dragonball/src/dbs_pci/src/msi.rs

```diff
@@ -0,0 +1,759 @@
+        match data.len() {
+            1 => match offset {
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() => {
+                    self.mask_bits = data[0] as u32 | (self.mask_bits & 0xffff_ff00);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 1 => {
+                    self.mask_bits = (data[0] as u32) << 8 | (self.mask_bits & 0xffff_00ff);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 2 => {
+                    self.mask_bits = (data[0] as u32) << 16 | (self.mask_bits & 0xff00_ffff);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 3 => {
+                    self.mask_bits = (data[0] as u32) << 24 | (self.mask_bits & 0x00ff_ffff);
+                }
+                _ => debug!("invalid offset"),
+            },
+            2 => {
+                let value = LittleEndian::read_u16(data);
+                match offset {
+                    MSI_MSG_CTL_OFFSET => {
+                        self.msg_ctl = (self.msg_ctl
+                            & !(MSI_CTL_ENABLE | MSI_CTL_MULTI_MSG_ENABLE))
+                            | value & (MSI_CTL_ENABLE | MSI_CTL_MULTI_MSG_ENABLE);
+                    }
+                    x if x == msg_data_offset => {
+                        self.msg_data = value;
+                    }
+                    x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() => {
+                        self.mask_bits = value as u32 | (self.mask_bits & 0xffff_0000);
+                    }
+                    x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 2 => {
+                        self.mask_bits = (value as u32) << 16 | (self.mask_bits & 0x0000_ffff);
+                    }
+                    _ => debug!("invalid offset"),
+                }
+            }
+            4 => {
+                let value = LittleEndian::read_u32(data);
+                match offset {
+                    0x0 => {
+                        self.msg_ctl = (self.msg_ctl
+                            & !(MSI_CTL_ENABLE | MSI_CTL_MULTI_MSG_ENABLE))
+                            | ((value >> 16) as u16 & (MSI_CTL_ENABLE | MSI_CTL_MULTI_MSG_ENABLE));
+                    }
+                    MSI_MSG_ADDR_LO_OFFSET => {
+                        self.msg_addr_lo = value & MSI_MSG_ADDR_LO_MASK;
+                    }
+                    x if addr_hi_offset.is_some() && x == addr_hi_offset.unwrap() => {
+                        self.msg_addr_hi = value;
+                    }
+                    x if x == msg_data_offset => {
+                        self.msg_data = value as u16;
+                    }
+                    x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() => {
+                        self.mask_bits = value;
+                    }
```

Comment:
Ditto.

Reply:
this is for writing bits according to pci spec. I have added a summarised comment above
"Support writing mask_bits, MSI control information, message data, message address."